### ISSUE TITLE
BUG(?): Array-priority for output wrapping seems wrong in ufuncs

### ISSUE THREAD

**Issue by seberg** [numpy/numpy#21280]
### Describe the issue:

UFuncs seem to not handle array priority wrapping quite right (at least compared to `np.get_array_wrap`).  That is, they simply skip NumPy arrays and scalars, but that is not necessarily right.
(The code should probably flag their existence, and return `NULL` manually when that makes sense)

### Reproduce the code example:

```python
import numpy as np
class myarr(np.ndarray):
    __array_priority__ = -1
a = np.arange(3)
b = np.arange(3).view(myarr)

# although priority is negative, `myarr` is used:
assert type(np.multiply(a, b)) is myarr
assert type(np.multiply(b, a)) is myarr

# But this is not true for the Python helper:
res = np.arange(3)
assert type(np.get_array_wrap(a, b)(res)) is np.ndarray
```

### NumPy/Python version information:

1.22.x/1.23.<dev>

**Comment by shivance** [comment]
Is this issue open to work on? I would love to contribute !

**Comment by seberg** [comment]
Yes of course, the tricky part here is probably mainly understanding the dispatching logic itself (and the fact that is on the C level).

**Comment by shivance** [comment]
@seberg I took a look in, just to confirm

```
# although priority is negative, `myarr` is used:
assert type(np.multiply(a, b)) is myarr
assert type(np.multiply(b, a)) is myarr
```
Here the type(np.multiply(a, b)) should be np.ndarray as it has higher priority than myarr which has negative priority ?

**Comment by seberg** [comment]
Yes, exactly.  For now, I will give you the pointer that you will need to look at the ufunc code which is in `numpy/core/src/umath/ufunc_object.c`, but only being imprecise, because I think it is probably a good learning experience to try to find out where this happens.

**Comment by shivance** [comment]
https://github.com/numpy/numpy/blob/5ecbc1445544bf04e6ce5f9dce6cda27c041660b/numpy/core/src/umath/ufunc_object.c#L4625-L4697

I think our problem lies somewhere in this block .

**Comment by shivance** [comment]
I did some digging up, and here are my thoughts
`replace_with_wrapped_result_and_return` is the function responsible for returning the final result with appropriate wrapper, here it goes.
Line Number 4656 in https://github.com/numpy/numpy/blob/5ecbc1445544bf04e6ce5f9dce6cda27c041660b/numpy/core/src/umath/ufunc_object.c#L4625-L4697 calls a function that is supposed to find appropriate wrap, It then goes to it's definition at line number 423 in same file. Then it jumps into method `_find_array_method` to determine the wrapping function given by the input arrays (docstrings say it all)

It then jumps to definition of `_find_array_method` , in which it loops over input array to get the wrapper with max priority
https://github.com/numpy/numpy/blob/5ecbc1445544bf04e6ce5f9dce6cda27c041660b/numpy/core/src/umath/ufunc_object.c#L217-L236

Now jumping into `Py_GetPriority` method we find

https://github.com/numpy/numpy/blob/c3a64134cd55bd6ed4689d601d5502c7cc0efbab/numpy/core/src/multiarray/multiarraymodule.c#L123-L152

Now, if we think, the only way myarray (with priority -1) will be wrapped if and only if the other arrays are having lesser priority.
here
`NPY_SCALAR_PRIORITY` has value -10000 and `NPY_PRIORITY` has value 0

@seberg what priority numpy array defined with `arange` method have ( how can we actually print priority of an array) ?

**Comment by seberg** [comment]
> @seberg what priority numpy array defined with arange method have ( how can we actually print priority of an array) ?

NumPy arrays have priority 0, which is `NPY_PRIORITY`.

**Comment by mattip** [comment]
See #17926

**Comment by shivance** [comment]
@mattip Thanks a lot. Learning how Python interact with C was one of the things that I always wanted to know <3

@seberg I am trying to understand how it is working, and turns out that I am in right direction, I put a breakpoint on `_find_array_method` which is hit, now when I'm trying to do `layout next` and debug line by line, I am finding assembly code, thus I am not able to understand how to use debugger here to check priority values in action.

![image](https://user-images.githubusercontent.com/51750587/199568390-4f902bcf-0a9f-4c92-8ede-da7235ce8ade.png)

**Comment by ganesh-k13** [comment]
I'm guessing you are using TUI? Did you try just printing the values you need and stepping through the code? That should provide helpful information to debug the issue.

**Comment by mattip** [comment]
By default, the C code is built with optimizations that remove all debugging information. You need to override this. One way to do that is to set `CFLAGS="-O0 -g"` before building.

**Comment by shivance** [comment]
@ganesh-k13 if you meant Terminal UI by TUI then yes, I'm using gdb in terminal
![image](https://user-images.githubusercontent.com/51750587/199904274-263eefef-9db0-4b8a-a2ab-f8e6fbe7de49.png)


I tried printing register values, but I don't think this is helpful at all as it's hard to interpret corresponding assembly of c for a student like me, but I am willing to learn. 

@mattip I launched gdb this time with `CFLAGS='O0 -g'` , is there a way I see the C code in layout window instead of compiled C objects and assembly registers?

Also I'm curious how numpy maintainers debug numpy as most of it's part is written in C?

**Comment by mattip** [comment]
Be sure you are setting the flags before building, this is what I usually do:
```
git clean -xfdd
CFLAGS="-O0 -g" gdb --args python runtests.py --python
```

**Comment by shivance** [comment]
@mattip great, now I am able to see which line it is breaking at, in console.

![image](https://user-images.githubusercontent.com/51750587/199929314-ebb522d8-667c-489b-bfe3-4f2ed6b710a8.png)

https://github.com/numpy/numpy/blob/5ecbc1445544bf04e6ce5f9dce6cda27c041660b/numpy/core/src/umath/ufunc_object.c#L217-L236

I want to confirm something here , the doc string of `n_method` says `If we have some methods defined, find the one of highest priority` . Please correct if my understanding is incorrect somewhere.

1. The goal of `_find_array_method` is to obtain the type of the resultant product of two numpy arrays
2. These python data types are referred as methods here
3. As we have defined a wrapper `myarr` , that makes the number of wrappers 2 (ndarray and myarr) 
4. To obtain the one with higher priority (ndarray : 0, myarr:-1) we wrote if block from line no 217 ownwards. 
5. If there is only one type, then it should return that

I had put breakpoints on line 221 and figured out that it is not executing, that means, numpy is inferring only one type (myarr) out of ndarray and myarr , and thus it is assigning output type myarr

Is there a way I can print PyObjects, (I only got numbers as output which are not understandable)

Thanks for the help

**Comment by seberg** [comment]
Sounds right enough for you to continue and try to find a way to fix it.

**Comment by shivance** [comment]
Okayy.
1. Is there a way I can print PyObjects, (I only got numbers as output which are not understandable) ?
2. Is my interpretation of n_methods correct ?

**Comment by seberg** [comment]
`PyObject_Print` can be used.  There are also gdb hooks that do some basic printing of python object also (I would have to look up what its called, many systems should make it easy to install/load).

The methods are the method that is attached to the object (slightly distinct, but right idea).

I do believe that it is an important skill to learn and press onward into the unknown, almost all code you have to touch is unknown to some degree (also for us).  So I would prefer to not give too many hints.

**Comment by shivance** [comment]
@ganesh-k13 @mattip I need a small help on part of debugger again. 
While debugging a code in python file, how to know which function is implemented in python and which one in C 
and if in C how to know corresponding C function location. I was trying to get a stacktrace of program, but not able to find anything online regarding gdb which gives stacktrace of underneath C code for line by line execution of Python program.

Thanks for the help.

**Comment by seberg** [comment]
Not sure it helps you, but https://devguide.python.org/advanced-tools/gdb/index.html has some information about the Python support for gdb.  That way you can get the Python backtrace when you are currently at a C breakpoint.

If you got a function that is not defined in Python (for example in `ipython` `??` doesn't show the code), things are a bit tricky.  Most of these go through this [struct](https://github.com/numpy/numpy/blob/main/numpy/core/src/multiarray/multiarraymodule.c#L4317), method [here](https://github.com/numpy/numpy/blob/main/numpy/core/src/multiarray/methods.c#L2847) although many call back into Python(!), ufuncs (the code you are editing) are auto-generated so a bit trickier.  But overall stepping through C is probably quite tedious compared to searching the code.
What are you looking for specifically, then I have an example to show you where to find it?

**Comment by shivance** [comment]
Hi, I did some research online, figured out a way to print python objects, and am in direction of localization of error.

But as I am debugging, gdb terminates at later point of time while stepping one by one 

It state that
```
4338    /usr/local/src/conda/python-3.9.13/Python/ceval.c: No such file or directory
```

Apparently it is using for system python but I am using conda python with numpy-dev environment, anyone came across this situation before, how did you set your python to virtual environment so that gdb looks at right location.
